linkerd/tls/route/src/lib.rs

```diff
@@ -0,0 +1,100 @@
+//! An TLS route matching library for Linkerd to support the TLSRoute
+//! Kubernetes Gateway API types.
+
+#![deny(rust_2018_idioms, clippy::disallowed_methods, clippy::disallowed_types)]
+#![forbid(unsafe_code)]
+
+use linkerd_tls::ServerName;
+use r#match::SessionMatch;
+use tracing::trace;
+
+#[derive(Clone, Debug, Hash, PartialEq, Eq)]
+pub struct SessionInfo {
+    pub sni: ServerName,
+}
+
+pub mod r#match;
+pub mod sni;
+#[cfg(test)]
+mod tests;
+
+pub use self::sni::{InvalidSni, MatchSni, SniMatch};
+
+/// Groups routing rules under a common set of SNIs.
+#[derive(Clone, Debug, Hash, PartialEq, Eq)]
+pub struct Route<P> {
+    /// A list of SNIs that this route applies to, to be matched against,
+    ///
+    /// If at least one match is specified, any match may apply for rules to applied.
+    /// When no SNI matches are present, all SNIs match.
+    pub snis: Vec<MatchSni>,
+
+    /// Must not be empty.
+    pub rules: Vec<Rule<P>>,
+}
+
+/// Policies for a given set of route matches.
+#[derive(Clone, Debug, Default, Hash, PartialEq, Eq)]
+pub struct Rule<P> {
+    /// A list of session matchers, *any* of which may apply.
+    ///
+    /// The "best" match is used when comparing rules.
+    pub matches: Vec<r#match::MatchSession>,
```

Comment:
So it seems that we have both SNI matching at the Route level and also the Rule level. If we're going to include MatchSni at the Route level, then we probably want MatchSession to be empty for now. Otherwise, we probably want to omit the sni matching at the Route level.

I don't have enough information to suggest one or the other. I don't really see any harm in it being defined at the Route level (since this matches the [gateway spec](https://github.com/kubernetes-sigs/gateway-api/blob/main/apis/v1alpha2/tlsroute_types.go#L87)). Given that the spec does not include any rule based matches currently, I'm not sure of a good reason to include SNI matches in rules.